### PR TITLE
[configtelemetry] add guidelines for each level of config telemetry

### LINKED ISSUE #10286: Describe configtelemetry levels
Prior to this package's stabilization, I would love it if there were more guidance on what each of the levels are for. Right now the documentation is very vague and I think the differences between normal & detailed are not specific enough. This is almost a flavor of #7543 where that metric could have been enabled only in the "detailed" configuration, however, reserving "detailed" that way wasn't an option because we put many metrics behind "detailed" that I would argue belong more in the "normal" or "basic" levels ([ex](https://github.com/open-telemetry/opentelemetry-collector/blob/main/processor/batchprocessor/metrics.go#L65-L67) of a detailed metric that IMO should be a "normal" metric).

_Originally posted by @jaronoff97 in https://github.com/open-telemetry/opentelemetry-collector/issues/10277#issuecomment-2142496499_

### PR DESCRIPTION
**Pull request by atoulme**
#### Description
Add to the package comment a set of guidelines for configtelemetry levels.

#### Link to tracking issue
Fixes #10286

### CODE REVIEW

**Comment by atoulme** [comment]
@jaronoff97 ptal

File: config/configtelemetry/doc.go
Comment (jaronoff97) [inline config/configtelemetry/doc.go:18]: Is there a maximum cardinality #  we should recommend here?

File: config/configtelemetry/doc.go
Comment (jaronoff97) [inline config/configtelemetry/doc.go:23]: How were these determined?

File: config/configtelemetry/doc.go
Comment (jaronoff97) [inline config/configtelemetry/doc.go:31]: what does control mean in this context?

File: config/configtelemetry/doc.go
Comment (jaronoff97) [inline config/configtelemetry/doc.go:34]: is this 20 individual meters or 20 metric points?

**Review by jaronoff97 — COMMENTED** [review]
A few thoughts, i like where this is heading.

**Comment by jaronoff97** [comment]
I do think in the future it would be neat if we had wrapper methods around otel APIs for recording telemetry for each level so authors don't need to keep track of these things.

File: config/configtelemetry/doc.go
Comment (atoulme) [reply config/configtelemetry/doc.go:23]: I made them up!

File: config/configtelemetry/doc.go
Comment (atoulme) [reply config/configtelemetry/doc.go:34]: 20 different metric time series, as in 20 combination of 20 metric names and dimension names values.

File: config/configtelemetry/doc.go
Comment (atoulme) [reply config/configtelemetry/doc.go:31]: There is a rationale for the cardinality and the author of the signal can explain the cardinality based on factors.
For example:
* The cardinality is linearly increasing with the number of Kafka topics processed
* The cardinality is correlated to the number of active clients

While this may incur a high cardinality, users can then filter based on it in some way or accumulate the signal by relying on this cardinality.

File: config/configtelemetry/doc.go
Comment (atoulme) [reply config/configtelemetry/doc.go:18]: sure?

File: config/configtelemetry/doc.go
Comment (jaronoff97) [reply config/configtelemetry/doc.go:31]: ah I see, thanks for adding that context. I think it would be useful to explain that a bit more in this comment – i.e. "it is acceptable at this level for cardinality to scale linearly with the monitored resources"

File: config/configtelemetry/doc.go
Comment (jaronoff97) [reply config/configtelemetry/doc.go:34]: Doesn't that mean a single metric that records the http status code would not fall into the normal level because the cardinality is >70?

Comment (jaronoff97) [reply config/configtelemetry/doc.go:34]: [http.server.request.duration](https://github.com/open-telemetry/semantic-conventions/blob/main/docs/http/http-metrics.md#metric-httpserverrequestduration)

Comment (atoulme) [reply config/configtelemetry/doc.go:34]: Checking, I think indeed that metric would be too verbose. You can create quite a bit of cardinality if you are running a server and any URI or method adds to cardinality. But say for the httpcheckreceiver, that might be fine because you use a http client to call just one endpoint, so cardinality is fixed.

**Comment by mx-psi** [comment]
@jaronoff97 Can you take another look?

File: config/configtelemetry/doc.go
Comment (jaronoff97) [reply config/configtelemetry/doc.go:34]: IMO given the metric is labeled "required" this should be covered by the normal telemetry settings.

**Comment by jaronoff97** [comment]
@mx-psi added another comment, i think other collector maintainers/developers should also look at this as well given it would potentially change what the collector is recommended to emit.

File: config/configtelemetry/doc.go
Comment (atoulme) [reply config/configtelemetry/doc.go:34]: ok, so let's relax this. Can you offer an amendment? Maybe we can set a max cardinality of 500 for now.
```suggestion
//   - A max cardinality (total possible combinations of dimension values) of 500.
```

File: config/configtelemetry/doc.go
Comment (atoulme) [reply config/configtelemetry/doc.go:31]:
```suggestion
// The signals associated with this level must control cardinality. It is acceptable at this level for cardinality to scale linearly with the monitored resources.
```

File: config/configtelemetry/doc.go
Comment (atoulme) [reply config/configtelemetry/doc.go:18]:
```suggestion
// The signals associated with this level must show low cardinality, the number of combinations of dimension values. The maximum cardinality allowed is set to 50.
```

File: config/configtelemetry/doc.go
Comment (jaronoff97) [reply config/configtelemetry/doc.go:34]: I think that's alright, I'd be interested in what enforcement looks like in the future given there's support for cardinality limits on instruments. https://github.com/open-telemetry/opentelemetry-go/blob/main/sdk/metric/internal/x/README.md#L11

Comment (atoulme) [reply config/configtelemetry/doc.go:34]: you probably get to write that part ;)

**Review by mwear — COMMENTED** [review]
I'm curious how configtelemetry level is meant to be used in the collector. Is the idea that it will be used to conditionally record telemetry throughout the codebase (are there any examples)? Have you considered using existing SDK mechanisms to limit telemetry such as samplers and views?

**Comment by atoulme** [comment]
> I'm curious how configtelemetry level is meant to be used in the collector. Is the idea that it will be used to conditionally record telemetry throughout the codebase (are there any examples)? Have you considered using existing SDK mechanisms to limit telemetry such as samplers and views?

you can look at obsreport to give you an idea of some of the bundled telemetry. We use the Go SDK for metrics and traces.

File: config/configtelemetry/doc.go
Comment (mx-psi) [inline config/configtelemetry/doc.go:16]:
```suggestion
// This is the default level recommended when running the Collector.
```

File: config/configtelemetry/doc.go
Comment (mx-psi) [inline config/configtelemetry/doc.go:18]: How are we going to enforce all the numbers in this doc?

Comment (atoulme) [reply config/configtelemetry/doc.go:18]: The issue says to offer guidance, not enforcement.

Before we go there, is this guidance any good? WDYT?

Comment (mx-psi) [reply config/configtelemetry/doc.go:18]: We don't need to implement enforcement on this PR or have a detailed plan of how to do it, but we need to know that these are reasonably easy to enforce in practice, otherwise the guidance is not actionable

Comment (atoulme) [reply config/configtelemetry/doc.go:18]: It looks reasonably easy to enforce in practice to me. The only tricky thing I guess is to calculate the cardinality of the metric, by computing the total set of combinations that can be formed based on the dimensions of the metric. Is that what needs to be detailed in there?

Comment (mx-psi) [reply config/configtelemetry/doc.go:18]: We should be able to measure the numbers easily, yes, my concern is more what are we going to do if we start deviating from these numbers. Do we get to a point where we just can't add more metrics at a certain level (since removing metrics is presumably a breaking change)? If a library that offers built-in instrumentation and starts producing many more metrics, what are we going to do?

To some extent this is a problem that we are going to have to solve regardless of what we put in here, but I think putting stringent requirements like these without having a good idea of how we are going we make our lives unnecessarily harder. I am tempted to just remove the numbers or make the numbers a rough guidance/approximation instead of a must-follow rule.

Comment (atoulme) [reply config/configtelemetry/doc.go:18]: > We should be able to measure the numbers easily, yes, my concern is more what are we going to do if we start deviating from these numbers. Do we get to a point where we just can't add more metrics at a certain level (since removing metrics is presumably a breaking change)? If a library that offers built-in instrumentation and starts producing many more metrics, what are we going to do?

We will need to reject the library in any case.

If we use say the Go SDK, and all of a sudden, it emits by default all sorts of metrics related to Go GC, this is in itself a change that can be considered a breaking change because it may break folks backends with high cardinality metrics.
> 
> To some extent this is a problem that we are going to have to solve regardless of what we put in here, but I think putting stringent requirements like these without having a good idea of how we are going we make our lives unnecessarily harder. I am tempted to just remove the numbers or make the numbers a rough guidance/approximation instead of a must-follow rule.

Yes, this whole doc is a guidance, I didn't mean for those numbers to feel such a hard rule. The numbers are what Pavol wanted to see, I am resistant also to provide absolute guidelines but offering numbers seemed to help enshrine some of the thinking.

Comment (mx-psi) [reply config/configtelemetry/doc.go:18]: > Yes, this whole doc is a guidance, I didn't mean for those numbers to feel such a hard rule

Can we made this more explicit in the text? (e.g. saying the limits are 'best effort' or adding a 'roughly' here or there to make it clear that this is a guideline but not a hard rule).

> offering numbers seemed to help enshrine some of the thinking.

Yup, I agree it makes the distinction between the levels clearer

File: config/configtelemetry/doc.go
Comment (djaglowski) [inline config/configtelemetry/doc.go:23]: I question whether logs should be included in this document at all. If they are, I would think guidance is given in terms of severity level, rather than telemetry level.

Log severity and sampling are well established ways to manage log volume and we already have configuration for them in the collector. I think that trying to bring in the concept of telemetry level in any way is going to be confusing at best.

File: config/configtelemetry/doc.go
Comment (djaglowski) [reply config/configtelemetry/doc.go:18]: Could we offer the numbers only as an example? e.g. "A typical component might have the following signals at this level."

IMO we're better off focusing the language on helping component developers make decisions about whether a metric or span should be included at a certain level, similar to how we might guide developers how to choose an appropriate log severity. (Also see my other comment about logs not fitting into this guidance.)

File: config/configtelemetry/doc.go
Comment (atoulme) [reply config/configtelemetry/doc.go:23]: I will remove them

File: config/configtelemetry/doc.go
Comment (atoulme) [reply config/configtelemetry/doc.go]: I have revised the text to take into account your considerations, please review.

**Comment by atoulme** [comment]
> focusing the language on helping component developers make decisions about whether a metric or span should be included at a certain level
That's a good insight and I have added a bit on the intent of the doc at the top. Please help make this better.

**Review by mx-psi — APPROVED** [review]

**Review by djaglowski — APPROVED** [review]
Thanks @atoulme. This looks good to me.

**Comment by atoulme** [comment]
@dmitryax reviewed with me and had feedback, we should wait before merging this. He explained that the "Basic" telemetry level is meant also to only be for core level libraries, which makes sense to me. It was discussed on an issue.

**Comment by dmitryax** [comment]
> @dmitryax reviewed with me and had feedback, we should wait before merging this. He explained that the "Basic" telemetry level is meant also to only be for core level libraries, which makes sense to me. It was discussed on an issue.

Here is the issue I was referencing https://github.com/open-telemetry/opentelemetry-collector/issues/7890

File: config/configtelemetry/doc.go
Comment (dmitryax) [inline config/configtelemetry/doc.go:21]: Currently, Normal is the default. Are we going to change that? I'd suggest not changing it but keeping this level lean instead

File: config/configtelemetry/doc.go
Comment (dmitryax) [inline config/configtelemetry/doc.go:18]:
```suggestion
// Signals associated with this level provide the essential coverage of the collector telemetry.
```

File: config/configtelemetry/doc.go
Comment (dmitryax) [inline config/configtelemetry/doc.go:19]:
```suggestion
// It should only be used for internal collector telemetry generated by the collector core API. Components SHOULD NOT use this level to report additional metrics.
```

File: config/configtelemetry/doc.go
Comment (dmitryax) [inline config/configtelemetry/doc.go:41]: How is this feasible? Collector is expected to support processing thousands of concurrent requests

File: config/configtelemetry/doc.go
Comment (dmitryax) [inline config/configtelemetry/doc.go:40]: Is it per monitored resource? If so it seems too much.

File: config/configtelemetry/doc.go
Comment (dmitryax) [inline config/configtelemetry/doc.go:27]: I don't think we should add this section, given that we don't expect components to emit at this level.

File: config/configtelemetry/doc.go
Comment (atoulme) [reply config/configtelemetry/doc.go]: fixed

File: config/configtelemetry/doc.go
Comment (atoulme) [reply config/configtelemetry/doc.go]: sounds good, removing

File: config/configtelemetry/doc.go
Comment (atoulme) [reply config/configtelemetry/doc.go:31]: true, does this clarify?
```suggestion
//   - At most 5 spans actively recording simultaneously per request
```

File: config/configtelemetry/doc.go
Comment (TylerHelmuth) [inline config/configtelemetry/doc.go:41]:
```suggestion
// The signals associated with this level may exhibit high cardinality and/or high dimensionality.
```

File: config/configtelemetry/doc.go
Comment (TylerHelmuth) [inline config/configtelemetry/doc.go:30]:
```suggestion
//     - A max cardinality (total possible combinations of dimension values) of 500.
```

File: config/configtelemetry/doc.go
Comment (TylerHelmuth) [inline config/configtelemetry/doc.go:31]:
```suggestion
//     - At most 5 spans actively recording simultaneously.
```

Comment (TylerHelmuth) [inline config/configtelemetry/doc.go:31]:
```suggestion
//     - At most 5 spans actively recording simultaneously.
```

**Review by TylerHelmuth — APPROVED** [review]

File: config/configtelemetry/doc.go
Comment (codeboten) [inline config/configtelemetry/doc.go:9]:
```suggestion
// When adopting a telemetry level, component authors are expected to rely on this guidance to
```

File: config/configtelemetry/doc.go
Comment (codeboten) [inline config/configtelemetry/doc.go:14]:
```suggestion
// No telemetry data is recorded.
```

File: config/configtelemetry/doc.go
Comment (codeboten) [inline config/configtelemetry/doc.go:18]:
```suggestion
// Telemetry associated with this level provides essential coverage of the collector telemetry.
```

File: config/configtelemetry/doc.go
Comment (codeboten) [inline config/configtelemetry/doc.go:19]:
```suggestion
// It should only be used for internal collector telemetry generated by the collector core API. Components outside of 
// the core API MUST NOT record additional telemetry at this level.
```

File: config/configtelemetry/doc.go
Comment (codeboten) [inline config/configtelemetry/doc.go:23]:
```suggestion
// Telemetry associated with this level provides complete coverage of the collector telemetry.
```

File: config/configtelemetry/doc.go
Comment (codeboten) [inline config/configtelemetry/doc.go:26]:
```suggestion
// Component authors using this telemetry level can use this guidance:
```

File: config/configtelemetry/doc.go
Comment (codeboten) [inline config/configtelemetry/doc.go:39]:
```suggestion
// Telemetry associated with this level provides complete coverage of the collector telemetry.
```

File: config/configtelemetry/doc.go
Comment (codeboten) [inline config/configtelemetry/doc.go:35]:
```suggestion
```

File: config/configtelemetry/doc.go
Comment (codeboten) [inline config/configtelemetry/doc.go:45]:
```suggestion
```

**Review by codeboten — COMMENTED** [review]
Some suggestions

File: config/configtelemetry/doc.go
Comment (atoulme) [inline config/configtelemetry/doc.go:31]:
```suggestion
//     - A max cardinality (total possible combinations of dimension values) for a given metric of at most 100.
```

File: config/configtelemetry/doc.go
Comment (atoulme) [inline config/configtelemetry/doc.go:32]:
```suggestion
//     - At most 5 spans actively recording simultaneously per active request.
```

File: config/configtelemetry/doc.go
Comment (atoulme) [reply config/configtelemetry/doc.go:31]: following up on @dmitryax comment above, lowering from 500 to 100.

File: config/configtelemetry/doc.go
Comment (atoulme) [reply config/configtelemetry/doc.go]: addressed below

**Review by jaronoff97 — APPROVED** [review]
Looks great, I appreciate that these are now recommendations. I think having basic remain the core telemetry also makes sense to me. Thank you for this.

**Review by codeboten — APPROVED** [review]

**Pull request merged**
